internal/controller/healthchecker.go

```diff
@@ -0,0 +1,161 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package controller
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type HealthChecker struct {
+	client.Reader    // intentional, we do not want to use the cache
+	Update           chan event.GenericEvent
+	Logger           log.Logger
+	NodeName         string
+	ApplyEmptyConfig EmptyConfigApplier
+	ReloadStatus     func()
+	ctx              context.Context
+	healthResult     string
+	healthResMutex   sync.Mutex
+	check            healthcheck
+	attempts         int
+	interval         time.Duration
+}
+
+type healthcheck func() error
+
+func (r *HealthChecker) HealthResult() string {
+	r.healthResMutex.Lock()
+	defer r.healthResMutex.Unlock()
+	return r.healthResult
+}
+
+type EmptyConfigApplier interface {
+	ApplyEmptyConfig() error
+}
+
+type hcEvent struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+}
+
+func (evt *hcEvent) DeepCopyObject() runtime.Object {
+	res := new(hcEvent)
+	res.Name = evt.Name
+	res.Namespace = evt.Namespace
+	return res
+}
+
+func NewHealthCheckEvent() event.GenericEvent {
+	evt := hcEvent{}
+	evt.Namespace = "healthcheck"
+	evt.Name = "check"
+	return event.GenericEvent{Object: &evt}
+}
+
+func NewHealthChecker(cl client.Reader, update chan event.GenericEvent, l log.Logger,
```

Comment:
Let's call this HealthCheckController because it's a controller (same for the file name)

Reply:
done